lldb/tools/lldb-dap/Transport.cpp

```diff
@@ -0,0 +1,146 @@
+//===-- Transport.cpp -----------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transport.h"
+#include "Protocol.h"
+#include "c++/v1/__system_error/error_code.h"
+#include "lldb/Utility/IOObject.h"
+#include "lldb/Utility/Status.h"
+#include "lldb/lldb-forward.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
+#include <string>
+#include <system_error>
+#include <utility>
+
+using namespace llvm;
+using namespace lldb;
+using namespace lldb_private;
+using namespace lldb_dap;
+using namespace lldb_dap::protocol;
+
+static Expected<std::string> ReadFull(IOObjectSP &descriptor, size_t length) {
```

Comment:
A (non-const) reference to a shared pointer usually not what you want. Since this function doesn't really need the shared-pointerness. I'd go with a reference to the underlying object instead.
```suggestion
static Expected<std::string> ReadFull(IOObject &descriptor, size_t length) {
```

Reply:
Tweaked to use a pointer to the `IOObject *` (which is abstract so I can't use a reference to the base class).

Reply:
That probably fine, but I am pretty sure that being abstract doesn't prevent you from [forming a reference](https://godbolt.org/z/Ejcxve5qq) to a class. You must have tried to copy the object somewhere. The main advantage of a reference is that it guarantees that the pointer is not null.